agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/ClientDecorator.java

```diff
@@ -15,10 +15,59 @@
  */
 package io.opentelemetry.auto.bootstrap.instrumentation.decorator;
 
+import io.grpc.Context;
+import io.opentelemetry.context.ContextUtils;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.DefaultSpan;
 import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.TracingContextUtils;
 
 public abstract class ClientDecorator extends BaseDecorator {
 
+  // Keeps track of the client span in a subtree corresponding to a client request.
+  // Visible for testing
+  static final Context.Key<Span> CONTEXT_CLIENT_SPAN_KEY =
+      Context.key("opentelemetry-trace-auto-client-span-key");
+
+  /**
+   * Creates a {@link Context} with the provided {@link Span} and sets it as the current {@link
+   * Context}
+   *
+   * @return a {@link Scope} that will restore the previous context. All callers of this method must
+   *     also call {@link Scope#close()} when this next context is no longer needed.
+   */
+  public static Scope currentContextWith(final Span clientSpan) {
+    return ContextUtils.withScopedContext(withSpan(clientSpan, Context.current()));
+  }
+
+  /** Returns a new {@link Context} forked from {@code context} with the {@link Span} set. */
+  public static Context withSpan(final Span clientSpan, final Context context) {
+    if (!clientSpan.getContext().isValid()) {
+      return TracingContextUtils.withSpan(clientSpan, context);
+    }
+    return TracingContextUtils.withSpan(
```

Comment:
I would extract common call to `TracingContextUtils.withSpan` from the `if` above.